Merge 3.1 branch

twiecki opened the pull request.

Commit:
Refactor Hamiltonian methods into single class

fonnesbeck committed:
Added mode argument to several step methods and advi to allow mode setting (e.g. nanguardmode) for Theano functions

fonnesbeck committed:
Fixed namespace bugs in mode attribute

fonnesbeck committed:
Reverted function in delta_logp to not accept mode argument

taku-y committed:
ENH KL-weighting

taku-y committed:
Fix bugs

ferrine committed:
ENH User model (#1525)

* Started to write Base class for pymc3.models

* mode `add_var` to public api

* Added some docstrings

* Added some docstrings

* added getitem and fixed a typo

* added assertion check

* added resolve var method

* decided not to add resolve method

* Added linear component

* Docs fix

* patsy's intercept is inited properly now

* refactored code

* updated docs

* added possibility to init coefficients with random variables

* added glm

* refactored api, fixed formula init

* refactored linear model, extended acceptable types

* moved useful matrix and labels creation to utils file

* code style

* removed redundant evaluation of shape

* refactored resolver for constructing matrix and labels

* changed error message

* changed signature of init

* simplified utils any_to_tensor_and_labels code

* tests for `any_to_tensor_and_labels`

* added docstring for `any_to_tensor_and_labels` util

* forgot to document return type in `any_to_tensor_and_labels`

* refactored code for dict

* dict tests fix(do not check labels there)

* added access to random vars of model

* added a shortcut for all variables so there is a unified way to get them

* added default priors for linear model

* update docs for linear

* refactored UserModel api, made it more similar to pm.Model class

* Lots of refactoring, tests for base class, more plain api design

* deleted unused module variable

* fixed some typos in docstring

* Refactored pm.Model class, now it is ready for inheritance

* Added documentation for Model class

* Small typo in docstring

* nested contains for treedict (needed for add_random_variable)

* More accurate duplicate implementation of treedict/treelist

* refactored treedict/treelist

* changed `__imul__` of treelist

* added `root` property and `isroot` indicator for base model

* protect `parent` and `model` attributes from violation

* travis' python2 did not fail on bad syntax(maybe it's too new), fixed

* decided not to use functools wrapper

 Unfortunately functools wrapper fails
 when decorating built-in methods so I
 need to fix that improper behaviour.
 Some bad but needed tricks were implemented

* Added models package to setup script

* Refactor utils

* Fix some typos in pm.model

ferrine committed:
ENH User model (#1525)

* Started to write Base class for pymc3.models

* mode `add_var` to public api

* Added some docstrings

* Added some docstrings

* added getitem and fixed a typo

* added assertion check

* added resolve var method

* decided not to add resolve method

* Added linear component

* Docs fix

* patsy's intercept is inited properly now

* refactored code

* updated docs

* added possibility to init coefficients with random variables

* added glm

* refactored api, fixed formula init

* refactored linear model, extended acceptable types

* moved useful matrix and labels creation to utils file

* code style

* removed redundant evaluation of shape

* refactored resolver for constructing matrix and labels

* changed error message

* changed signature of init

* simplified utils any_to_tensor_and_labels code

* tests for `any_to_tensor_and_labels`

* added docstring for `any_to_tensor_and_labels` util

* forgot to document return type in `any_to_tensor_and_labels`

* refactored code for dict

* dict tests fix(do not check labels there)

* added access to random vars of model

* added a shortcut for all variables so there is a unified way to get them

* added default priors for linear model

* update docs for linear

* refactored UserModel api, made it more similar to pm.Model class

* Lots of refactoring, tests for base class, more plain api design

* deleted unused module variable

* fixed some typos in docstring

* Refactored pm.Model class, now it is ready for inheritance

* Added documentation for Model class

* Small typo in docstring

* nested contains for treedict (needed for add_random_variable)

* More accurate duplicate implementation of treedict/treelist

* refactored treedict/treelist

* changed `__imul__` of treelist

* added `root` property and `isroot` indicator for base model

* protect `parent` and `model` attributes from violation

* travis' python2 did not fail on bad syntax(maybe it's too new), fixed

* decided not to use functools wrapper

 Unfortunately functools wrapper fails
 when decorating built-in methods so I
 need to fix that improper behaviour.
 Some bad but needed tricks were implemented

* Added models package to setup script

* Refactor utils

* Fix some typos in pm.model

taku-y committed:
Remove unnecessary comments

Commit:
Reformat docs

taku-y committed:
Fix typo

taku-y committed:
Minor fixes

taku-y committed:
Check transformed RVs using hasattr

ferrine committed:
shape problem when sampling

ferrine committed:
tests passed

taku-y committed:
Update conv-vae notebook

ferrine committed:
refactored GARCH and added Mv(Gaussian/StudentT)RandomWalk (#1603)

* refactored GARCH and added Mv(Gaussian/StudentT)RandomWalk

* refactored garch logp

* added docs

* fix typo

* even more typos

* better description for tau

Commit:
Refactor and make more descriptive

fonnesbeck committed:
Generalized Bound to work with discrete variables

fonnesbeck committed:
Fixed distributions/__init__.py to reflect change to Bound location

twiecki committed:
MAINT Replace discrete RV check.

MAINT Replace discrete RV check.

twiecki committed:
MAINT Use theano GammaLn and Psi functions. (#1628)

fonnesbeck committed:
Added EdwardModel and example

fonnesbeck committed:
Bounded now inherits from Distribution

fonnesbeck committed:
Added newline at end of edward_beta_bernoulli

twiecki committed:
Remove static reference to theano.config.floatX. Use .astype(). Addresses @ferrine's comments.

Remove static reference to theano.config.floatX. Use .astype(). Addresses @ferrine's comments.

twiecki committed:
ENH Add floatX() to theanof and make advi use it.

twiecki committed:
Remove nan_. Add doc-string.

twiecki committed:
ENH default to asarray if int or float is passed.

fonnesbeck committed:
Added newline to end of edward.py

Commit:
Bump to python 3.6

twiecki commented:
Had to rebase, hopefully I didn't mess it up.

twiecki commented on pymc3/distributions/distribution.py:
```diff
@@ -2,17 +2,21 @@
 import theano.tensor as tt
 from theano import function
 
+from . import transforms
```
This creates a circular dependency. Not sure how it worked before. @fonnesbeck ?

ColCarroll replied:
that was introduced https://github.com/pymc-devs/pymc3/pull/1625, and tests passed (https://coveralls.io/builds/9484082)

ColCarroll replied:
gosh, we should do a real fix for this, but I think there's a silly one, too

twiecki replied:
What's the silly one? :)

ColCarroll replied:
Tried/failed on my own branch -- it is funny that tests pass on python3, which made me think it is just import order.

fredcallaway committed:
allow one-sided patsy formula in linear_component

fredcallaway committed:
fix bug inspecting patsy formula

ferrine committed:
fix cyclic import

twiecki committed:
BLD Add missing external to install.

twiecki committed:
BLD Typo externals -> external.

twiecki committed:
BLD Add step_methods.hmc to install.

twiecki committed:
MAINT Add multivariate random walks to imports.

aloctavodia committed:
foresplot: hide transformed variables by default

twiecki committed:
Merge pull request #1668 from ColCarroll/bump_test_versions

Bump to python 3.6

ferrine commented:
in 7f9542531d2ca31339bbeaec8d8bf477266d69eb I fix cyclic import, you can do the same

twiecki committed:
DOC Started release-notes for 3.1.

twiecki commented:
Thanks @ferrine. Not sure why the exact sample test for HMC is failing here. @ColCarroll, ideas?

ColCarroll commented:
I feel like I've seen this exact failure before -- in the travis logs the traces look exactly the same (though I guess only the first 23 entries are actually the same).  I can look more closely later, but probably need to print the full traces.

twiecki commented:
Most of them differ by E-20, but I saw one differing on order of .1.

fonnesbeck commented:
If we are going to add a `pymc3.models` submodule, then we need to reconcile that with `pymc3.model`. We should not have both. 

Also, we have things like `timeseries`, which is just another type of model, but it is not in `models`. This requires more discussion and consideration, but I don't think having a generic `models` submodule is the best way to go.

ferrine commented:
Can we try `setuptools.find_packages()` in setup script after fixing hmc? There were some issues with missing modules, they are easy to fix but` setuptools` suggest a better solution.

ColCarroll commented:
Ah, update on HMC -- I looked into it for a while, and took a larger sample from a working SHA and HEAD.  The larger samples had the same behavior: they'd be equal for a while, then all of a sudden be different, the converge again, though there was no readily apparent pattern in how different they were.  It is interesting because 

- even though HMC shares machinery with NUTS, only HMC is failing,
-  the momentum samples remain exactly the same
- somehow the traces reconverge

The algorithm still seems fine -- my working theory is that the calls to the numpy random number generator are getting evaluated in a different order sometimes, or some other function is making a draw from the RNG.  I'm not against replacing the test with the current trace, but I'll probably keep looking at this if only because I'm intrigued now.

ColCarroll commented:
In particular, I think the other call to an RNG is for the metropolis acceptance step, so it *seems* like that must be it...

twiecki commented:
@ColCarroll If that's the issue, does it suggest that the Metropolis accept does not use the correct seed? How do we fix it?

ColCarroll commented:
It'll take more testing to figure this out -- everything uses `numpy.random`, so the seed is set right.  What I suspect happens is, roughly:

On master:
```
for step in steps:
    hmc_sample_momentum()
    hmc_metropolis_accept()
    # occasionally
    something_else_draw_sample()
```

In current branch:

On master:
```
for step in steps:
    hmc_sample_momentum()
    # occasionally
    something_else_draw_sample()
    hmc_metropolis_accept()
```

This would explain why it is usually in sync, but goes into and out of sync.

twiecki commented:
OK, so if we change the stored samples we should be good?

ColCarroll commented:
Yeah, seems like.  The statistical check on HMC still passes.  Will rest
easier if I knew why it changed, though.

On Tue, Jan 17, 2017, 10:05 AM Thomas Wiecki <notifications@github.com>
wrote:

> OK, so if we change the stored samples we should be good?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/pymc-devs/pymc3/pull/1658#issuecomment-273193670>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACMHEHdNfoDVphTKxhQrF2sAOnduDKalks5rTNipgaJpZM4Lebbh>
> .
>

twiecki commented:
@ferrine Can you push the update? Maybe we can decouple the easy resting from making travis happy. At least you looked deep enough that there aren't clear bugs.

ferrine committed:
add `find_packages()` (#1675)

travis is happy, the problem is still in hmc

ferrine commented:
I've merged my PR with `find_packages` into 3.1 but it did not appear here. Hmm, I did something wrong?
UPD: it appeared

ColCarroll commented:
wowza -- after some close checking, there is a transposed negative sign in the hmc sampler.  Since hmc rarely rejects samples, the differences were very close.  I'll add that change in.

ColCarroll committed:
Fix hmc bug

Was computing negative energy for HMC.

ColCarroll committed:
Fix test_step.py

The Metropolis tuning algorithm needed just a little nudge in the right direction.

twiecki commented:
@ColCarroll Wow, those tests really paid off here!

twiecki commented:
OK, I'm going to merge this so that we get out of this limbo.

fonnesbeck commented:
Nice work guys!
